### PR TITLE
Blog: Authenticating using selenium

### PR DESCRIPTION
**Pull request by psiinon**
Depends on https://github.com/zaproxy/community-scripts/pull/297

TODO:

- [x] Wait for https://github.com/zaproxy/community-scripts/pull/297 to be merged
- [x] Add one or more images
- [x] Update the date

As you can see the blog post isnt ready, but it does help explain how the scripts in [#297](https://github.com/zaproxy/community-scripts/pull/297) work so may well aid that review.

Signed-off-by: Simon Bennetts <psiinon@gmail.com>

### CODE REVIEW

**Comment by psiinon** [comment]
Screenshot 2023-01-31 at 17-24-33 Authenticating Using Selenium

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
+---
+title: "Authenticating Using Selenium"
+summary: >
+  How to configure ZAP to handling complex authentication using Selenium.
```
Comment (thc202) [inline site/content/blog/2023-02-01-authenticating-using-selenium/index.md:4]: `handle`?

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+We can in fact configure ZAP to authenticate to Juice shop _without_ using Selenium, but we will ignore that option here :wink:.
+
+To automate ZAP we are going to use the [Automation Framework](/docs/automate/automation-framework/) (AF) - this is a powerful and flexible framework ideally suited to this task.
+
+### The Solution
+
+The solution is quite complex, and requires several scripts to make it work.
+
+The full set of scripts and instructions on how to run then are available at
```
Comment (thc202) [inline site/content/blog/2023-02-01-authenticating-using-selenium/index.md:73]: `them`?

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+The first time it runs it create a new proxy for the authentication requests - this is how ZAP can tell that these requests are for authentication.
+
+The proxy is stored in a [global script variable](/docs/desktop/addons/script-console/#global-variables) so that we do not keep trying to recreate it.
+
+The script then launches a browser using the [Selenium add-on](/docs/desktop/addons/selenium/) specifying that it should use the new proxy.
+
+It actually makes 2 requests so that we avoid the warning modals that Juice Shop uses - these are very hard to get rid of in any other way.
+
+If you need to use Selenium to authenticate to your apps then its the commands that control the browser in this script that you will need to change to match your app.
```
Comment (thc202) [inline site/content/blog/2023-02-01-authenticating-using-selenium/index.md:120]: `it's`?

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+The [JuiceShopSelenium.js](https://github.com/zaproxy/community-scripts/tree/main/af-plans/juiceshop-selenium-auth/JuiceShopSelenium.js) script is run whenever ZAP launches a browser. This will happen when the AJAX Spider and DOM XSS rule run, but also when we launch a browser via the authentication script.
+
+The script also logs in to Juice Shop in the same way as the authentication script. This sets up the client side authentication state correctly - without this the AJAX Spider would not be able to explore authenticated content.
+
+If your apps set client side authentication state then you will need to authenticate in a similar way.
+
+#### The Session Management Script
+
+The [JuiceShopSession.js](https://github.com/zaproxy/community-scripts/tree/main/af-plans/juiceshop-selenium-auth/JuiceShopSession.js) script is run when direct request are made by ZAP.
```
Comment (thc202) [inline site/content/blog/2023-02-01-authenticating-using-selenium/index.md:144]: `requests`

**Comment by psiinon** [comment]
Image added and date changed to tomorrow.
Ready for review...

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+It updates the requests with the authentication token and cookie.
+
+It will not run when the AJAX Spider runs as we are running that 'unauthenticated' as it logs in to Juice Shop when the browsers are launched.
+
+#### The Reset Script
+
+The [JuiceShopReset.js](https://github.com/zaproxy/community-scripts/tree/main/af-plans/juiceshop-selenium-auth/JuiceShopReset.js) script is not actually used when automating ZAP. However it is very useful when you are testing and debugging a set up like this using the ZAP desktop.
+
+The script stops the proxy started by the authentication script, clears the session tokesn stored in global script variables and resets the authentication states for all of the users.
```
Comment (thc202) [inline site/content/blog/2023-02-01-authenticating-using-selenium/index.md:154]: `tokens`

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+### Sanity Checks
+
+You have managed to configure ZAP to authenticate to your app and test it effectively.
+
+How do you know that this set up will keep working?
+
+As you will have seen the above plans use a combination of standard ZAP stats and customs ones which are maintained by the scripts.
+
+However the AF supports more [Job Tests](/docs/desktop/addons/automation-framework/tests/) so you can do things like stopping long running jobs early based on statistic thresholds and checking that ZAP finds specific URLs (and optionally specific content them). You could use the latter to ensure ZAP finds URLs that it should only be able to find if authentication is working.
```
Comment (thc202) [inline site/content/blog/2023-02-01-authenticating-using-selenium/index.md:164]: `in them`?

**Comment by psiinon** [comment]
Thanks - fixed those

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+In addition to that it runs the standard spider authenticated and the AJAX Spider unauthenticated - ZAP is not used to authenticate the AJAX Spider because the browsers are logged in when they are launched.
+
+In both cases a set of statistics are checked to make sure that authentication is really working - both standard 
+[ZAP stats](/docs/internal-statistics/) are used as are statistics maintained by the scripts below.
+
+#### The Authentication Script
+
+The [JuiceShopAuthentication.js](https://github.com/zaproxy/community-scripts/tree/main/af-plans/juiceshop-selenium-auth/JuiceShopAuthentication.js) script is run when ZAP detects that a direct request is not currently authenticated.
+
+The first time it runs it create a new proxy for the authentication requests - this is how ZAP can tell that these requests are for authentication.
```
Comment (kingthorin) [inline site/content/blog/2023-02-01-authenticating-using-selenium/index.md:112]:
```suggestion
The first time it runs it creates a new proxy for the authentication requests - this is how ZAP can tell that these requests are for authentication.
```

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+The [JuiceShopSession.js](https://github.com/zaproxy/community-scripts/tree/main/af-plans/juiceshop-selenium-auth/JuiceShopSession.js) script is run when direct requests are made by ZAP. 
+
+It updates the requests with the authentication token and cookie.
+
+It will not run when the AJAX Spider runs as we are running that 'unauthenticated' as it logs in to Juice Shop when the browsers are launched.
+
+#### The Reset Script
+
+The [JuiceShopReset.js](https://github.com/zaproxy/community-scripts/tree/main/af-plans/juiceshop-selenium-auth/JuiceShopReset.js) script is not actually used when automating ZAP. However it is very useful when you are testing and debugging a set up like this using the ZAP desktop.
```
Comment (kingthorin) [inline site/content/blog/2023-02-01-authenticating-using-selenium/index.md:152]:
```suggestion
The [JuiceShopReset.js](https://github.com/zaproxy/community-scripts/tree/main/af-plans/juiceshop-selenium-auth/JuiceShopReset.js) script is not actually used when automating ZAP. However, it is very useful when you are testing and debugging a set up like this using the ZAP desktop.
```

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+The script stops the proxy started by the authentication script, clears the session tokens stored in global script variables and resets the authentication states for all of the users.
```
Comment (kingthorin) [inline site/content/blog/2023-02-01-authenticating-using-selenium/index.md:154]:
```suggestion
The script stops the proxy started by the authentication script, clears the session tokens stored in global script variables, and resets the authentication states for all of the users.
```

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+### Sanity Checks
+
+You have managed to configure ZAP to authenticate to your app and test it effectively.
+
+How do you know that this set up will keep working?
+
+As you will have seen the above plans use a combination of standard ZAP stats and customs ones which are maintained by the scripts.
+
+However the AF supports more [Job Tests](/docs/desktop/addons/automation-framework/tests/) so you can do things like stopping long running jobs early based on statistic thresholds and checking that ZAP finds specific URLs (and optionally specific content in them). You could use the latter to ensure ZAP finds URLs that it should only be able to find if authentication is working.
```
Comment (kingthorin) [inline site/content/blog/2023-02-01-authenticating-using-selenium/index.md:164]:
```suggestion
However, the AF supports more [Job Tests](/docs/desktop/addons/automation-framework/tests/) so you can do things like stopping long running jobs early based on statistic thresholds and checking that ZAP finds specific URLs (and optionally specific content in them). You could use the latter to ensure ZAP finds URLs that it should only be able to find if authentication is working.
```

File: site/content/blog/2023-02-01-authenticating-using-selenium/index.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+We can in fact configure ZAP to authenticate to Juice shop _without_ using Selenium, but we will ignore that option here :wink:.
+
+To automate ZAP we are going to use the [Automation Framework](/docs/automate/automation-framework/) (AF) - this is a powerful and flexible framework ideally suited to this task.
+
+### The Solution
+
+The solution is quite complex, and requires several scripts to make it work.
+
+The full set of scripts and instructions on how to run them are available at
+https://github.com/zaproxy/community-scripts/tree/js-sel-auth/af-plans/juiceshop-selenium-auth
```
[the remark this answers is not in the document]
Comment (psiinon) [reply site/content/blog/2023-02-01-authenticating-using-selenium/index.md:74]: It will linkify this, but I'll tweak it anyway.
Oh, and fix the link!

**Comment by psiinon** [comment]
Thnaks, hopefully fixed those

**Review by kingthorin — APPROVED** [review]
Thanks

**Pull request merged**